src/io/delimited.cr

```diff
@@ -0,0 +1,124 @@
+module IO
+  # An IO that wraps another IO, and only reads up to the beginning of a
+  # specified delimiter.
+  #
+  # This is useful for exposing part of an underlying stream to a client.
+  #
+  # ```
+  # io = MemoryIO.new "abc||123"
+  # delimited = IO::Delimited.new(io, read_delimiter: "||")
+  #
+  # delimited.gets_to_end # => "abc"
+  # delimited.gets_to_end # => nil
+  # io.gets_to_end        # => "123"
+  # ```
+  class Delimited
+    include IO
+
+    # If `sync_close` is true, closing this IO will close the underlying IO.
+    property? sync_close
+
+    getter read_delimiter
+    getter? closed : Bool
+
+    @delimiter_buffer : Bytes
+    @active_delimiter_buffer : Bytes?
```

Comment:
I don't know why it's so much faster. Some guesses:
1. The original code had more checks. Not only the nil checks, but then setting the slice to nil or increment it by one depending on some condition (and this is done on every iteration)
2. Unions are codegen'd in LLVM as pair: type id and opaque bytes. These bytes are casted when accessed/stored, so maybe LLVM doesn't optimize that well. But now I'm thinking that we could represent nilable types of just two types (Nil | T) better, as type id + T, because Nil has no real value. I'll have to try that and see if the performance is regained. If so, it will be an extremely useful general optimization, but I don't know if that's the reason.